### PR TITLE
Add validation for `answer_count` in when clauses

### PR DESCRIPTION
**Pull request by benhowes**
This PR adds syntax validation for routing_rules which make use of the number of answers to a repeated question.

This PR:
1. Adds JSON schema defintions
2. Adds sanity checking which cannot be easily caught by the schema validation
3. Adds a test to ensure incorrect inputs are found

### CODE REVIEW

File: app/validation/validator.py [hunk 1]
```diff
@@ -195,11 +195,20 @@ def validate_routing_on_answer_options(self, block, answer):
...
         # Validates any answer id in a when clause exists within the schema
+        answer_reference_id_fields = ('id', 'answer_count')
+
         for when in when_clause:
-            if 'id' in when:
-                if when['id'] not in existing_ids:
-                    return Validator._error_message('The answer id - {} in the "when" clause for {} does not exist'
-                                                    .format(when['id'], referenced_id))
+            # either of the reference fields may be in use
+            id_ref_key = [x for x in answer_reference_id_fields if x in when].pop()
```
Comment (nealedj) [inline app/validation/validator.py]: I had to double take on this pattern. Could `next` do the same thing?

```
id_ref_key = next((x for x in answer_reference_id_fields if x in when), None)
```

File: tests/test_schema_validation.py [hunk 2]
```diff
@@ -209,6 +209,19 @@ def test_invalid_survey_id_whitespace(self):
 
         self.assertEquals(errors.get('message'), "'lms ' does not match '^[0-9a-z]+$'")
 
+    def test_invalid_routing_when_answer_count(self):
+        """Asserts that invalid `when` routing_rules are caught for `answer_count`"""
+        file = 'schemas/test_invalid_routing_when_answer_count.json'
```
Comment (nealedj) [inline tests/test_schema_validation.py]: Nitpick - this overrides the builtin `file` function.

**Comment by benhowes** [comment]
@nealedj I'd not accounted for the use of the `meta` key in when clauses. I've updated it to not mind if there are no id references in the when clause.

Running `./scripts/test_schemas.sh --local` from the survey runner now gives me: `137 Passed - 0 Failed`

File: tests/test_schema_validation.py (hunk 2 above)
Comment (benhowes) [reply tests/test_schema_validation.py]: I just followed the convention in from existing tests - I can change them all in this PR, or do that in a separate refactor one?

**Comment by benhowes** [comment]
Ready review again @nealedj

File: app/validation/validator.py [hunk 3]
```diff
@@ -200,8 +200,16 @@ def validate_when_rule(when_clause, existing_ids, referenced_id):
...
 
         for when in when_clause:
-            # either of the reference fields may be in use
-            id_ref_key = [x for x in answer_reference_id_fields if x in when].pop()
+            # either of the reference fields may be in use, however it will
+            # never be the case that both are defined since this is handled by
+            # the JSON schema check
+            present_id_ref_keys = [x for x in answer_reference_id_fields if x in when]
+
+            if not len(present_id_ref_keys):
```
Comment (nealedj) [inline app/validation/validator.py]: Another minor one - unnecessary call to `len`. Empty lists evaluate to `False` anyway.

Not blocking.

Comment (benhowes) [reply app/validation/validator.py]: Made the change in any case

**Comment by nealedj** [comment]
I think the above is because #61 was merged.

**Comment by benhowes** [comment]
@nealedj this seems to be resolved after https://github.com/ONSdigital/eq-survey-runner/pull/1675 has been merged. I've rebased this on master.

**Comment by kierenbeckett** [comment]
You'll need to rebase eq-survey-runner a too

**Comment by kierenbeckett** [comment]
LGTM, you'll need to squash and sign your commit before I can merge it

**Comment by benhowes** [comment]
@kierenbeckett Squashed

**Pull request merged**